// @well-known-components/nats-component/dist/test-component.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.createLocalNatsComponent = void 0;
const mitt_1 = __importDefault(require("mitt"));
/**
 * Create a local NATS component, for testing purposes
 * @public
 */
async function createLocalNatsComponent() {
    const callbacks = new Map();
    const events = (0, mitt_1.default)();
    function publish(topic, data) {
        callbacks.forEach((cbs, pattern) => {
            const sPattern = pattern.split(".");
            const sTopic = topic.split(".");
            if (sPattern.length !== sTopic.length) {
                return;
            }
            for (let i = 0; i < sTopic.length; i++) {
                if (sPattern[i] !== "*" && sPattern[i] !== sTopic[i]) {
                    return;
                }
            }
            for (const cb of cbs) {
                cb(null, { subject: topic, data });
            }
        });
    }
    function subscribe(pattern, cb) {
        const cbs = callbacks.get(pattern) || new Set();
        cbs.add(cb);
        callbacks.set(pattern, cbs);
        return {
            unsubscribe: () => {
                cbs.delete(cb);
            },
        };
    }
    async function start() {
        events.emit("connected");
    }
    async function stop() { }
    return {
        publish,
        subscribe,
        start,
        stop,
        events,
    };
}
exports.createLocalNatsComponent = createLocalNatsComponent;
